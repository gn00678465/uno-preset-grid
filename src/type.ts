export interface GridOptions {
  columns?: number
  breakpoints?: Record<string, string>
  gutter?: number
  piece?: number
  /**
   * Prefix for CSS variables.
   *
   * @default 'un-'
   * */
  variablePrefix?: string
  /**
   * 1rem = n px
   * @default 16
   */
  baseFontSize?: number
  /**
   * Length units
   * px | rem
   * @default 'px
   */
  lengthUnit?: 'px' | 'rem'
}
