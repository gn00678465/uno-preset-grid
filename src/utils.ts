const remRE = /(-?[.\d]+)rem/g

const pxRE = /(-?[.\d]+)px/g

/**
 * 將 rem 轉換為 px
 * @param size 
 * @param baseFontSize 
 * @returns 
 */
export function convertRemToPx(size: string, baseFontSize = 16) {
  if (typeof size === 'string' && remRE.test(size)){
    return size.replace(remRE, (_, p1) => `${p1 * baseFontSize}px`)
  }
  return size
}

/**
 * 將 px 轉換為 rem
 * @param size 
 * @param baseFontSize 
 * @returns 
 */
export function convertPxToRem(size: string, baseFontSize = 16) {
  if (typeof size === 'string' && pxRE.test(size)){
    return size.replace(pxRE, (_, p1) => `${p1 / baseFontSize}rem`)
  }
  return size
}

/**
 * 轉換長度單位
 * @param size 
 * @param unit 
 * @param baseFontSize 
 * @returns 
 */
export function convertLenUnit(size: string, unit: 'px' | 'rem', baseFontSize = 16) {
  if (unit === 'px') return convertRemToPx(size, baseFontSize)
  if (unit === 'rem') return convertPxToRem(size, baseFontSize)
  return size
}

export function convertBreakpointSize(size: string | number, baseFontSize = 16) {
  if (typeof size === 'number') return size
  const _px = convertLenUnit(size, 'px', baseFontSize)
  if (_px.endsWith('px')) return parseInt(_px.replace('px', ''))
  return 0
}

export function getBreakpointEntries(breakpoints: Record<string, string>, baseFontSize = 16) {
  return Object.entries(breakpoints)
  .map(([k, v]) => [k, convertBreakpointSize(v, baseFontSize)])
  .filter(([, v]) => !!v)
  .sort(([, a], [, b]) => (b as number) - (a as number)) as([string, number])[]
}