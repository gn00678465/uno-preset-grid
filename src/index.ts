import { toEscapedSelector as e, type Preset } from 'unocss'
import type { Theme } from "unocss/preset-mini"
import type { GridOptions } from "./type"
import { mobileFirstBreakpointSort, convertLenUnit } from './utils'

export function presetGrid (options: GridOptions = {}): Preset {
  const {
    columns = 12,
    gutter = 24,
    variablePrefix = 'un-',
    baseFontSize = 16,
    lengthUnit = 'px'
  } = options
  // Default breakpoints
  const maxWidth = options?.maxWidth ?? {
    sm: '640px',
    md: '768px',
    lg: '1024px',
    xl: '1280px',
    '2xl': '1536px'
  }
  const breakpoints = options?.breakpoints ?? {
    sm: '640px',
    md: '768px',
    lg: '1024px',
    xl: '1280px',
    '2xl': '1536px'
  }

  const gutterXVar = `--${variablePrefix}gutter-x`
  const gutterYVar = `--${variablePrefix}gutter-y`
  const gutterXLen = `${gutter}px`

  return {
    name: 'uno-preset-grid',
    rules: [
      /** container */
      [
        new RegExp(`^flex-container$`),
        ([], { generator, rawSelector }) => {
          const _breakpoints = (generator?.userConfig?.theme as Theme)?.breakpoints ?? breakpoints
          const _breakpointEntries = mobileFirstBreakpointSort(_breakpoints)
          const _selector = e(rawSelector)

          let containerStyle = `
          ${_selector},
          ${_selector}-fluid {
            ${gutterXVar}: ${convertLenUnit(gutterXLen, lengthUnit, baseFontSize)};
            ${gutterYVar}: 0;
            width: 100%;
            padding-right: calc(var(${gutterXVar}) * 0.5);
            padding-left: calc(var(${gutterXVar}) * 0.5);
            margin-right: auto;
            margin-left: auto;
          }
          `
          _breakpointEntries.forEach(([breakpoint, value]) => {
            const width = maxWidth[breakpoint] || `${value}px`
            containerStyle += `
            @media (min-width: ${value}px) {
              ${_selector} {
                max-width: ${width};
              }
            }
            `
          })

          

          return containerStyle
        },
        { autocomplete: ['flex-container', 'flex-container-fluid'] }
      ],
      /** row */
      [
        new RegExp(`^row$`),
        function*(_, { symbols }) {
          yield {
            [gutterXVar]: gutterXLen,
            [gutterYVar]: '0',
            display: 'flex',
            'flex-wrap': 'wrap',
            'margin-top': `calc(-1 * var(${gutterYVar}))`,
            'margin-right': `calc(-.5 * var(${gutterXVar}))`,
            'margin-left': `calc(-.5 * var(${gutterXVar}))`,
          }
          yield {
            [symbols.selector]: selector => `${selector} > *`,
            'flex-shrink': 0,
            width: '100%',
            'max-width': '100%',
            'padding-right': `calc(var(${gutterXVar}) * .5)`,
            'padding-left': `calc(var(${gutterXVar}) * .5)`,
            'margin-top': `var(${gutterYVar})`
          }
        }
      ],
      /** col */
      [
        new RegExp(`^(\\w+)?:?col-?(\\d*)$`),
        function*([_, breakpoint, size], { symbols, generator }) {
          const _breakpoints = (generator?.userConfig?.theme as Theme)?.breakpoints ?? breakpoints
          if (!breakpoint) {
            if (!size) {
              yield {
                flex: '1 0 0%'
              }
            }
            if (!isNaN(parseInt(size)) && parseInt(size) > 0 && parseInt(size) <= columns) {
              yield {
                [symbols.selector]: selector => `${selector}`,
                flex: '0 0 auto',
                width: `${(parseInt(size) / columns) * 100}%`
              }
            }
          }
          if (breakpoint && breakpoint in _breakpoints) {
            if (!isNaN(parseInt(size)) && parseInt(size) > 0 && parseInt(size) <= columns) {
              yield {
                [symbols.parent]: `@media (min-width: ${_breakpoints[breakpoint]})`,
                [symbols.selector]: selector => `${selector}`,
                flex: '0 0 auto',
                width: `${(parseInt(size) / columns) * 100}%`
              }
            }
            if (!size) {
              yield {
                [symbols.parent]: `@media (min-width: ${_breakpoints[breakpoint]})`,
                [symbols.selector]: selector => `${selector}`,
                flex: '1 0 0%'
              }
            }
          }
          return undefined
        },
        { autocomplete: ['col', 'col-<num>'] }
      ],
      /** gutter */
      [
        new RegExp(`^g([xy])?-(\\d+)$`),
        ([, dim, size]) => {
          let gutterObject: { [key: string]: string } = {}
          if (dim !== "y") gutterObject[gutterXVar] = `${parseInt(size) * 16}px`
          if (dim !== "x") gutterObject[gutterYVar] = `${parseInt(size) * 16}px`
          return gutterObject
        },
        { autocomplete: ["g-<num>", "gx-<num>", "gy-<num>"] }
      ]
    ],
    postprocess: (util) => {
      util.entries.forEach((i) => {
        const [property, value] = i
        if (property.startsWith(`--${variablePrefix}`) && typeof value === 'string') {
          i[1] = convertLenUnit(value, lengthUnit, baseFontSize)
        }
      })
    }
  }
}
