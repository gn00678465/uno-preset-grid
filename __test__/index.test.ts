import { createGenerator } from 'unocss'
import { describe, expect, it } from 'vitest'
import { presetGrid } from '../src/index'

describe('preset-grid', () => {
  it('generate flex-container', async () => {
    const uno = await createGenerator({
      presets: [
        presetGrid()
      ],
    })
    const { css: containerCss } = await uno.generate([
      'flex-container',
      'flex-container-fluid',
      'lg:flex-container'
    ])
    expect(containerCss).toMatchSnapshot()
  })

  it('different container piece', async () => {
    const uno = await createGenerator({
      presets: [
        presetGrid({
          piece: 32
        })
      ],
    })
    const { css: noPreflightCSS } = await uno.generate([
      'flex-container',
      'flex-container-fluid',
      'lg:flex-container'
    ])
    expect(noPreflightCSS).toMatchSnapshot()
  })

  it('generate row', async () => {
    const uno = await createGenerator({
      presets: [
        presetGrid()
      ],
    })
    const { css: noPreflightCSS } = await uno.generate('row')
    expect(noPreflightCSS).toMatchSnapshot()
  })

  it('generate gutter', async () => {
    const uno = await createGenerator({
      presets: [
        presetGrid()
      ],
    })
    const { css: noPreflightCSS } = await uno.generate(['gy-1', 'g-12', 'gx-6', 'gx-0', 'gy-0'])
    expect(noPreflightCSS).toMatchSnapshot()
  })

  it('generate col', async () => {
    const uno = await createGenerator({
      presets: [
        presetGrid()
      ],
    })
    const { css: noPreflightCSS } = await uno.generate([
      'col',
      'col-1',
      'col-6',
      'col-12',
      'md:col-1',
      'md:col',
      'sm:col-12',
      'lg:col-6',
      'col-13',
      'xl:col-13'
    ])
    expect(noPreflightCSS).toMatchSnapshot()
  })

  it('generate gutter', async () => {
    const uno = await createGenerator({
      presets: [
        presetGrid()
      ],
    })
    const { css: noPreflightCSS } = await uno.generate([
      'g-1',
      'gx-3',
      'gy-6'
    ])
    expect(noPreflightCSS).toMatchSnapshot()
  })

  it('different length unit', async () => {
    const uno = await createGenerator({
      presets: [
        presetGrid({
          lengthUnit: 'rem',
          baseFontSize: 16
        })
      ],
    })
    const { css: noPreflightCSS } = await uno.generate([
      'flex-container',
      'flex-container-fluid',
      'row',
      'col',
      'col-1',
      'col-12',
      'md:col',
      'sm:col-12',
      'lg:col-6',
      'g-1',
      'gx-3',
      'gy-6'
    ])
    expect(noPreflightCSS).toMatchSnapshot()
  })

  it('different prefix', async () => {
    const uno = await createGenerator({
      presets: [
        presetGrid({
          lengthUnit: 'rem',
          baseFontSize: 16,
          variablePrefix: 'test-'
        })
      ],
    })
    const { css: noPreflightCSS } = await uno.generate([
      'flex-container',
      'flex-container-fluid',
      'row',
      'col',
      'col-1',
      'col-12',
      'md:col',
      'sm:col-12',
      'lg:col-6',
      'g-1',
      'gx-3',
      'gy-6'
    ])
    expect(noPreflightCSS).toMatchSnapshot()
  })

  it('with other breakpoints', async () => {
    const uno = await createGenerator({
      presets: [
        presetGrid({
          breakpoints: {
            sm: '576px',
            md: '768px',
            lg: '992px',
            xl: '1200px',
            xxl: '1400px'
          }
        })
      ],
    })

    const { css: noPreflightCSS } = await uno.generate([
      'flex-container',
      'row',
      'col',
      'col-1',
      'col-12',
      'sm:col-2',
      'md:col-3',
      'lg:col-4',
      'xl:col-5',
      'col-13'
    ])

    expect(noPreflightCSS).toMatchSnapshot()
  })
})