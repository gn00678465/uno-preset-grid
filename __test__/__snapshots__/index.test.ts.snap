// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`preset-grid > different container piece 1`] = `
"/* layer: default */
.flex-container,
.flex-container-fluid{--un-gutter-x:24px;--un-gutter-y:0;width:100%;padding-right:calc(var(--un-gutter-x) * 0.5);padding-left:calc(var(--un-gutter-x) * 0.5);margin-right:auto;margin-left:auto;}
@media (min-width: 1024px){
.flex-container{max-width:calc(1024px - 32px);}
}
@media (min-width: 1280px){
.flex-container{max-width:calc(1280px - 32px);}
}
@media (min-width: 1536px){
.flex-container{max-width:calc(1536px - 32px);}
}
@media (min-width: 640px){
.flex-container{max-width:calc(640px - 32px);}
}
@media (min-width: 768px){
.flex-container{max-width:calc(768px - 32px);}
}"
`;

exports[`preset-grid > different length unit 1`] = `
"/* layer: default */
.flex-container,
.flex-container-fluid{--un-gutter-x:1.5rem;--un-gutter-y:0;width:100%;padding-right:calc(var(--un-gutter-x) * 0.5);padding-left:calc(var(--un-gutter-x) * 0.5);margin-right:auto;margin-left:auto;}
.row{--un-gutter-x:1.5rem;--un-gutter-y:0;display:flex;flex-wrap:wrap;margin-top:calc(-1 * var(--un-gutter-y));margin-right:calc(-.5 * var(--un-gutter-x));margin-left:calc(-.5 * var(--un-gutter-x));}
.row > *{flex-shrink:0;width:100%;max-width:100%;padding-right:calc(var(--un-gutter-x) * .5);padding-left:calc(var(--un-gutter-x) * .5);margin-top:var(--un-gutter-y);}
.col{flex:1 0 0%;}
.col-1{flex:0 0 auto;width:8.333333333333332%;}
.col-12{flex:0 0 auto;width:100%;}
.g-1{--un-gutter-x:1rem;--un-gutter-y:1rem;}
.gx-3{--un-gutter-x:3rem;}
.gy-6{--un-gutter-y:6rem;}
@media (min-width: 1024px){
.flex-container{max-width:calc(1024px - 60px);}
.lg\\:col-6{flex:0 0 auto;width:50%;}
}
@media (min-width: 1280px){
.flex-container{max-width:calc(1280px - 60px);}
}
@media (min-width: 1536px){
.flex-container{max-width:calc(1536px - 60px);}
}
@media (min-width: 640px){
.flex-container{max-width:calc(640px - 60px);}
.sm\\:col-12{flex:0 0 auto;width:100%;}
}
@media (min-width: 768px){
.flex-container{max-width:calc(768px - 60px);}
.md\\:col{flex:1 0 0%;}
}"
`;

exports[`preset-grid > different prefix 1`] = `
"/* layer: default */
.flex-container,
.flex-container-fluid{--test-gutter-x:1.5rem;--test-gutter-y:0;width:100%;padding-right:calc(var(--test-gutter-x) * 0.5);padding-left:calc(var(--test-gutter-x) * 0.5);margin-right:auto;margin-left:auto;}
.row{--test-gutter-x:1.5rem;--test-gutter-y:0;display:flex;flex-wrap:wrap;margin-top:calc(-1 * var(--test-gutter-y));margin-right:calc(-.5 * var(--test-gutter-x));margin-left:calc(-.5 * var(--test-gutter-x));}
.row > *{flex-shrink:0;width:100%;max-width:100%;padding-right:calc(var(--test-gutter-x) * .5);padding-left:calc(var(--test-gutter-x) * .5);margin-top:var(--test-gutter-y);}
.col{flex:1 0 0%;}
.col-1{flex:0 0 auto;width:8.333333333333332%;}
.col-12{flex:0 0 auto;width:100%;}
.g-1{--test-gutter-x:1rem;--test-gutter-y:1rem;}
.gx-3{--test-gutter-x:3rem;}
.gy-6{--test-gutter-y:6rem;}
@media (min-width: 1024px){
.flex-container{max-width:calc(1024px - 60px);}
.lg\\:col-6{flex:0 0 auto;width:50%;}
}
@media (min-width: 1280px){
.flex-container{max-width:calc(1280px - 60px);}
}
@media (min-width: 1536px){
.flex-container{max-width:calc(1536px - 60px);}
}
@media (min-width: 640px){
.flex-container{max-width:calc(640px - 60px);}
.sm\\:col-12{flex:0 0 auto;width:100%;}
}
@media (min-width: 768px){
.flex-container{max-width:calc(768px - 60px);}
.md\\:col{flex:1 0 0%;}
}"
`;

exports[`preset-grid > generate col 1`] = `
"/* layer: default */
.col{flex:1 0 0%;}
.col-1{flex:0 0 auto;width:8.333333333333332%;}
.col-12{flex:0 0 auto;width:100%;}
.col-6{flex:0 0 auto;width:50%;}
@media (min-width: 1024px){
.lg\\:col-6{flex:0 0 auto;width:50%;}
}
@media (min-width: 640px){
.sm\\:col-12{flex:0 0 auto;width:100%;}
}
@media (min-width: 768px){
.md\\:col{flex:1 0 0%;}
.md\\:col-1{flex:0 0 auto;width:8.333333333333332%;}
}"
`;

exports[`preset-grid > generate flex-container 1`] = `
"/* layer: default */
.flex-container,
.flex-container-fluid{--un-gutter-x:24px;--un-gutter-y:0;width:100%;padding-right:calc(var(--un-gutter-x) * 0.5);padding-left:calc(var(--un-gutter-x) * 0.5);margin-right:auto;margin-left:auto;}
@media (min-width: 1024px){
.flex-container{max-width:calc(1024px - 60px);}
}
@media (min-width: 1280px){
.flex-container{max-width:calc(1280px - 60px);}
}
@media (min-width: 1536px){
.flex-container{max-width:calc(1536px - 60px);}
}
@media (min-width: 640px){
.flex-container{max-width:calc(640px - 60px);}
}
@media (min-width: 768px){
.flex-container{max-width:calc(768px - 60px);}
}"
`;

exports[`preset-grid > generate gutter 1`] = `
"/* layer: default */
.g-12{--un-gutter-x:192px;--un-gutter-y:192px;}
.gx-0{--un-gutter-x:0px;}
.gx-6{--un-gutter-x:96px;}
.gy-0{--un-gutter-y:0px;}
.gy-1{--un-gutter-y:16px;}"
`;

exports[`preset-grid > generate gutter 2`] = `
"/* layer: default */
.g-1{--un-gutter-x:16px;--un-gutter-y:16px;}
.gx-3{--un-gutter-x:48px;}
.gy-6{--un-gutter-y:96px;}"
`;

exports[`preset-grid > generate row 1`] = `
"/* layer: default */
.row{--un-gutter-x:24px;--un-gutter-y:0;display:flex;flex-wrap:wrap;margin-top:calc(-1 * var(--un-gutter-y));margin-right:calc(-.5 * var(--un-gutter-x));margin-left:calc(-.5 * var(--un-gutter-x));}
.row > *{flex-shrink:0;width:100%;max-width:100%;padding-right:calc(var(--un-gutter-x) * .5);padding-left:calc(var(--un-gutter-x) * .5);margin-top:var(--un-gutter-y);}"
`;

exports[`preset-grid > with other breakpoints 1`] = `
"/* layer: default */
.flex-container,
.flex-container-fluid{--un-gutter-x:24px;--un-gutter-y:0;width:100%;padding-right:calc(var(--un-gutter-x) * 0.5);padding-left:calc(var(--un-gutter-x) * 0.5);margin-right:auto;margin-left:auto;}
.row{--un-gutter-x:24px;--un-gutter-y:0;display:flex;flex-wrap:wrap;margin-top:calc(-1 * var(--un-gutter-y));margin-right:calc(-.5 * var(--un-gutter-x));margin-left:calc(-.5 * var(--un-gutter-x));}
.row > *{flex-shrink:0;width:100%;max-width:100%;padding-right:calc(var(--un-gutter-x) * .5);padding-left:calc(var(--un-gutter-x) * .5);margin-top:var(--un-gutter-y);}
.col{flex:1 0 0%;}
.col-1{flex:0 0 auto;width:8.333333333333332%;}
.col-12{flex:0 0 auto;width:100%;}
@media (min-width: 1200px){
.flex-container{max-width:calc(1200px - 60px);}
.xl\\:col-5{flex:0 0 auto;width:41.66666666666667%;}
}
@media (min-width: 1400px){
.flex-container{max-width:calc(1400px - 60px);}
}
@media (min-width: 576px){
.flex-container{max-width:calc(576px - 60px);}
.sm\\:col-2{flex:0 0 auto;width:16.666666666666664%;}
}
@media (min-width: 768px){
.flex-container{max-width:calc(768px - 60px);}
.md\\:col-3{flex:0 0 auto;width:25%;}
}
@media (min-width: 992px){
.flex-container{max-width:calc(992px - 60px);}
.lg\\:col-4{flex:0 0 auto;width:33.33333333333333%;}
}"
`;
